server/src/main/java/com/thoughtworks/go/server/service/AuthorizationExtensionCacheService.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2019 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.service;
+
+import com.google.common.base.Ticker;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.thoughtworks.go.config.PluginProfile;
+import com.thoughtworks.go.config.PluginRoleConfig;
+import com.thoughtworks.go.config.SecurityAuthConfig;
+import com.thoughtworks.go.plugin.access.authorization.AuthorizationExtension;
+import com.thoughtworks.go.plugin.domain.authorization.AuthenticationResponse;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Component
+public class AuthorizationExtensionCacheService {
+    private final int CACHE_EXPIRY_IN_MINUTES = 30;
+
+    private Cache<String, Boolean> isValidUserCache;
+    private Cache<String, AuthenticationResponse> getUserRolesCache;
+    private final AuthorizationExtension authorizationExtension;
+    private final String IS_VALID_USER_CACHE_KEY = "is_valid_user_request_for";
+
+    public AuthorizationExtensionCacheService(AuthorizationExtension authorizationExtension, Ticker ticker) {
+        this.authorizationExtension = authorizationExtension;
+        isValidUserCache = CacheBuilder.newBuilder()
+                .ticker(ticker).expireAfterWrite(CACHE_EXPIRY_IN_MINUTES, TimeUnit.MINUTES).build();
+        getUserRolesCache = CacheBuilder.newBuilder()
+                .ticker(ticker).expireAfterWrite(CACHE_EXPIRY_IN_MINUTES, TimeUnit.MINUTES).build();
```

Comment:
What is the default max capacity of the cache? Should we consider setting a max size on the number of items in cache?

Reply:
@ketan -- there is no maximum size on the cache. Not sure if we want to add a max cache size. As the number of elements in the cache can never be greater than the total number of access tokens.
And the cache will be invalidated after 30 mins, so no eviction too is required.